implementations/rust/ockam/ockam_node_test_attribute/src/parser.rs

```diff
@@ -0,0 +1,128 @@
+use proc_macro::TokenStream;
+
+use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
+use quote::{quote, quote_spanned, ToTokens};
+use syn::token;
+
+use crate::args;
+use crate::args::Args;
+use crate::hygiene;
+
+pub(crate) fn parse_macro(
+    input: syn::ItemFn,
+    args: syn::AttributeArgs,
+    is_test: bool,
+) -> Result<TokenStream, syn::Error> {
+    let (input, ctx_pat) = hygiene::input(input, is_test)?;
+    let args = args::parse(args, is_test)?;
+    let last_stmt_end_span = {
+        let mut last_stmt = input
+            .block
+            .stmts
+            .last()
+            .map(ToTokens::into_token_stream)
+            .unwrap_or_default()
+            .into_iter();
+        // `Span` on stable Rust has a limitation that only points to the first
+        // token, not the whole tokens. We can work around this limitation by
+        // using the first/last span of the tokens like
+        // `syn::Error::new_spanned` does.
+        let start = last_stmt.next().map_or_else(Span::call_site, |t| t.span());
+        last_stmt.last().map_or(start, |t| t.span())
+    };
+    let ctx_ident = &ctx_pat.ident;
+    // If the input function doesn't use a mutable Context, it won't be able to stop it.
+    let ctx_stop_stmt = if ctx_pat.mutability.is_none() {
```

Comment:
This was just to make it easier to write the tests somehow, but I agree it moves the complexity to the macro, which is bad. Also, without this, the tests will be more explicit regarding how the context is used, so let's remove this part.

Reply:
Should we then force a mutable borrow in the `node_test` attribute? Currently, there is no limitation in this regard.

Reply:
The thing for test functions is that for happy-path tests this becomes easier to handle because you can just stop the context at the end of the test. But if the test panics, how do we stop the context? This should be done by the macro, otherwise we are forcing the user to wrap the test in another function so they can stop the context if it failed... which is weird. Makes sense?

Reply:
But it's important that test is not hanging because of panic/failure. Timeout should deal with that just fine